contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,423 @@
+
+        // get current token balances in pool
+        uint256 xBal = IERC20(token0).balanceOf(address(this));
+        uint256 yBal = IERC20(token1).balanceOf(address(this));
+
+        // convert balances to wad for liquidity calcs adjusted for protocol fees
+        uint256 xWad = _amountWad(xBal - protocolFees0, true);
+        uint256 yWad = _amountWad(yBal - protocolFees1, false);
+
+        // get D balance before add liquidity
+        uint256 _d = _D(xWad, yWad);
+
+        // transfer amounts in decimals in
+        if (amounts[0] > 0) IERC20(token0).safeTransferFrom(msg.sender, address(this), amounts[0]);
+        if (amounts[1] > 0) IERC20(token1).safeTransferFrom(msg.sender, address(this), amounts[1]);
+
+        // update amounts for actual transfer amount in case of fees on transfer
+        amounts[0] = IERC20(token0).balanceOf(address(this)) - xBal;
+        amounts[1] = IERC20(token1).balanceOf(address(this)) - yBal;
+
+        // convert amounts to wad for calcs
+        uint256 amount0Wad = _amountWad(amounts[0], true);
+        uint256 amount1Wad = _amountWad(amounts[1], false);
+
+        // balances after transfer
+        xWad += amount0Wad;
+        yWad += amount1Wad;
+
+        // calc diff with new D value
+        minted_ = _D(xWad, yWad) - _d;
+        _d += minted_;
+
+        uint256[] memory fees = new uint256[](2);
+        emit AddLiquidity(msg.sender, amounts, fees, _d, _d);
+    }
+
+    /// @notice Removes liquidity from pool
+    /// @param amounts The token amounts to remove in token decimals
+    /// @param deadline The deadline before which liquidity must be removed
+    // TODO: test w protocol fees
+    function removeLiquidity(uint256[] memory amounts, uint256 deadline)
+        external
+        onlyOwner
+        deadlineCheck(deadline)
+        returns (uint256 burned_)
+    {
+        require(amounts.length == 2, "invalid amounts");
+
+        // get current token balances in pool
+        uint256 xBal = IERC20(token0).balanceOf(address(this));
+        uint256 yBal = IERC20(token1).balanceOf(address(this));
+        require(amounts[0] + protocolFees0 <= xBal, "dx > max");
+        require(amounts[1] + protocolFees1 <= yBal, "dy > max");
+
+        // convert balances to wad for liquidity calcs adjusted for protocol fees
+        uint256 xWad = _amountWad(xBal - protocolFees0, true);
+        uint256 yWad = _amountWad(yBal - protocolFees1, false);
+
+        // get D balance before remove liquidity
+        uint256 _d = _D(xWad, yWad);
+
+        // transfer amounts out
+        if (amounts[0] > 0) IERC20(token0).safeTransfer(msg.sender, amounts[0]);
+        if (amounts[1] > 0) IERC20(token1).safeTransfer(msg.sender, amounts[1]);
+
+        // update amounts for actual transfer amount in case of fees on transfer
+        amounts[0] = xBal - IERC20(token0).balanceOf(address(this));
+        amounts[1] = yBal - IERC20(token1).balanceOf(address(this));
```

Comment:
I think this check could be omitted, as the potential fee could only decrease the received amount, but the amount sent is always what was specified in `.transfer()`. And we should be fine emitting event with "amounts sent" here.